File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/iteration/MarshallerBuilder.scala
package org.infinispan.server.hotrod.iteration

import org.infinispan.commons.marshall.Marshaller
import org.infinispan.commons.marshall.jboss.GenericJBossMarshaller
import org.infinispan.filter.KeyValueFilterConverter

import scala.util.Try

/**
 * @author gustavonalle
 * @since 8.0
 */
object MarshallerBuilder {
   def toClass[K, V, C](filter: IterationFilter[K, V, C]) = filter.marshaller.map(_.getClass).orNull

   def fromClass[K, V, C](marshallerClass: Option[Class[Marshaller]], filter: Option[KeyValueFilterConverter[K, V, C]]): Marshaller = {
      val withClassLoaderCtor = for {
         f <- filter
         m <- marshallerClass
         c <- Try(m.getConstructor(classOf[ClassLoader])).toOption
      } yield c.newInstance(filter.getClass.getClassLoader)

      withClassLoaderCtor.getOrElse {
         withEmptyCtor(marshallerClass).getOrElse(genericFromInstance(filter))
      }
   }

   private def withEmptyCtor(marshallerClass: Option[Class[Marshaller]]) = marshallerClass.map(_.newInstance())

   def genericFromInstance(instance: Option[AnyRef]): Marshaller = {
      new GenericJBossMarshaller(instance.map(_.getClass.getClassLoader).orNull)
   }
}

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/iteration/MarshallerBuilder.java
package org.infinispan.server.hotrod.iteration;

import java.lang.reflect.InvocationTargetException;
import java.util.Optional;

import org.infinispan.commons.CacheException;
import org.infinispan.commons.marshall.Marshaller;
import org.infinispan.commons.marshall.jboss.GenericJBossMarshaller;
import org.infinispan.filter.KeyValueFilterConverter;

/**
 * @author gustavonalle
 * @since 8.0
 */
class MarshallerBuilder {
   static <K, V, C> Class<?> toClass(IterationFilter<K, V, C> filter) {
      return filter.marshaller.map(Object::getClass).orElse(null);
   }

   static Marshaller fromClass(Optional<Class<Marshaller>> marshallerClass, Optional<KeyValueFilterConverter> filter) {
      return filter.flatMap(f -> marshallerClass.map(m -> {
         try {
            return m.getConstructor(ClassLoader.class);
         } catch (NoSuchMethodException e) {
            throw new CacheException(e);
         }
      }).map(c -> {
         try {
            return c.newInstance(f.getClass().getClassLoader());
         } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new CacheException(e);
         }
      })).orElseGet(() -> marshallerClass.map(c -> {
         try {
            return c.newInstance();
         } catch (InstantiationException | IllegalAccessException e) {
            throw new CacheException(e);
         }
      }).orElseGet(() -> genericFromInstance(filter)));
   }

   static Marshaller genericFromInstance(Optional<?> instance) {
      return new GenericJBossMarshaller(instance.map(i -> i.getClass().getClassLoader()).orElse(null));
   }
}
